### PR TITLE
Setup list to ts

### PR DESCRIPTION
**Pull request by foxriver76**
And some fixes for States and objects types which came up

### CODE REVIEW

File: packages/db-objects-redis/src/lib/objects/objectsInRedisClient.ts [hunk 1]
```diff
@@ -2536,10 +2536,14 @@ export class ObjectsInRedisClient {
 
     // -------------- OBJECT FUNCTIONS -------------------------------------------
     // No callback provided by user, we return a Promise
-    private _subscribe(pattern: string, asUser: boolean): Promise<void>;
-    // Callback provided by user, we call callback
-    private _subscribe(pattern: string, asUser: boolean, callback?: ioBroker.ErrorCallback): void;
-    private _subscribe(pattern: string, asUser: boolean, callback?: ioBroker.ErrorCallback): void | Promise<void> {
+    private _subscribe(pattern: string | string[], asUser: boolean): Promise<void>;
+    // Callback provided by user, we call callback|
+    private _subscribe(pattern: string | string[], asUser: boolean, callback?: ioBroker.ErrorCallback): void;
```
Comment (AlCalzone) [inline packages/db-objects-redis/src/lib/objects/objectsInRedisClient.ts:2541]:
```suggestion
    private _subscribe(pattern: string | string[], asUser: boolean, callback: ioBroker.ErrorCallback): void;
```
optional (missing) would match the previous overload

File: packages/db-objects-redis/src/lib/objects/objectsInRedisClient.ts [hunk 2]
```diff
@@ -4431,6 +4435,19 @@ export class ObjectsInRedisClient {
         return result;
     }
 
+    // getObjectList is called without options
+    getObjectList(
+        params: ioBroker.GetObjectListParams,
+        callback?: ioBroker.GetObjectListCallback
+    ): void | Promise<ioBroker.CallbackReturnTypeOf<ioBroker.GetObjectListCallback>>;
```
Comment (AlCalzone) [inline packages/db-objects-redis/src/lib/objects/objectsInRedisClient.ts:4442]: maybe distinguish the return type here (promise or not) depending on the callback arg like you did for _subscribe?

File: packages/cli/src/lib/setup/setupList.ts [hunk 3]
```diff
@@ -0,0 +1,916 @@
...
+    alive?: boolean;
+    enabled?: boolean;
+    disabled?: boolean;
+    port?: number;
+    ssl?: boolean;
+    ip?: string;
+}
+
+export class List {
+    private config: Record<string, any>;
```
Comment (AlCalzone) [inline packages/cli/src/lib/setup/setupList.ts:37]: We could think about adding a proper interface which describes the shape of the system config. Probably not part of this PR though

File: packages/cli/src/lib/setup/setupList.ts [hunk 4]
```diff
@@ -0,0 +1,916 @@
...
+    private objects: ObjectsRedisClient;
+    private states: StatesRedisClient;
+    private readonly processExit: (exitCode?: number) => void;
+
+    constructor(options: Record<string, any>) {
```
Comment (AlCalzone) [inline packages/cli/src/lib/setup/setupList.ts:42]: There is probably a better type for `options`?

File: packages/cli/src/lib/setup/setupList.ts [hunk 5]
```diff
@@ -0,0 +1,916 @@
...
+        return result;
+    }
+
+    showFileHeader(adapter: string): void {
+        adapter && console.log(`\n-- ${adapter} --`);
+        console.log('   Modified at  |Permission|     User     |     Group    | Size | Name');
+        console.log('----------------+----------+--------------+--------------+------+---------');
+    }
+
+    showFile(adapter: string, path: string, file: Record<string, any>): void {
```
Comment (AlCalzone) [inline packages/cli/src/lib/setup/setupList.ts:85]: Is there a better type for `file`?

File: packages/cli/src/lib/setup/setupList.ts [hunk 6]
```diff
@@ -0,0 +1,916 @@
...
+            text += (file.isDir ? 'd' : '-') + List._perm2str(file.acl.permissions || 0);
+            let owner = file.acl.owner;
+            if (owner) {
+                owner = owner.substring(12);
+                if (owner.length < 15) {
+                    owner = new Array(15 - owner.length).join(' ') + owner;
+                }
+            } else {
+                owner = new Array(15).join(' ');
+            }
```
Comment (AlCalzone) [inline packages/cli/src/lib/setup/setupList.ts:109]: This looks like it could be replaced with string.padStart/padEnd

**Review by AlCalzone — COMMENTED** [review]
Looks more or less fine from a TS perspective. There is a bunch of old-school code though, which could be modernized, like padding strings, using async/await in a for loop instead of recursively calling callbacks.

File: packages/db-objects-redis/src/lib/objects/objectsInRedisClient.ts (hunk 1 above)
Comment (foxriver76) [reply packages/db-objects-redis/src/lib/objects/objectsInRedisClient.ts:2541]: The problem is that, internally the inference is not working without this type, because we may pass undefined explicitly
Bildschirmfoto von 2022-05-11 09-11-36

Comment (foxriver76) [reply packages/db-objects-redis/src/lib/objects/objectsInRedisClient.ts:2541]: I can try to solve it via `conditionals` so if `undefined` Promise, else callback

File: packages/db-objects-redis/src/lib/objects/objectsInRedisClient.ts [hunk 7]
```diff
@@ -2537,8 +2537,12 @@ export class ObjectsInRedisClient {
...
     // No callback provided by user, we return a Promise
     private _subscribe(pattern: string | string[], asUser: boolean): Promise<void>;
-    // Callback provided by user, we call callback|
-    private _subscribe(pattern: string | string[], asUser: boolean, callback?: ioBroker.ErrorCallback): void;
+    // Callback provided by user, we call callback
+    private _subscribe<T extends ioBroker.ErrorCallback>(
+        pattern: string | string[],
+        asUser: boolean,
+        callback?: T
+    ): T extends ioBroker.ErrorCallback ? void : Promise<void>;
```
Comment (AlCalzone) [inline packages/db-objects-redis/src/lib/objects/objectsInRedisClient.ts:2545]: I think the first overload can be removed now (assuming the conditional type is ok)

File: packages/cli/src/lib/setup/setupList.ts [hunk 8]
```diff
@@ -33,13 +33,19 @@ interface FlagObject {
     ip?: string;
 }
 
+interface ListOptions {
```
Comment (AlCalzone) [inline packages/cli/src/lib/setup/setupList.ts:36]: Just a nit: When I started reworking some of the setup classes to CLIxyz, part of the intent was to make it clear these are CLI commands.
I think it would be good to reflect this in the options interface name, e.g.
```suggestion
interface CLIListOptions {
```
In the long run we could then merge the CLI and setup variants of the implementations and have them named consistently.
